File: wooyoo/Programmers/2019_카카오_개발자_겨울_인턴십/PR_크레인_인형_뽑기_게임.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>

using namespace std;

// 프로그래머즈 2019 카카오 개발자 겨울 인턴십 크레인 인형뽑기 게임
// 간단한 구현 문제


int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    int arr_size = board.size();

    stack<int> bucket;

    for (int i = 0; i < moves.size(); i++) {

        int column = moves[i] - 1;

        for (int j = 0; j < arr_size; j++) {

            int doll = board[j][column];


            if (doll != 0) {
                cout << doll << endl;
                if (!bucket.empty() && bucket.top() == doll) {
                    bucket.pop();
                    answer++;
                } else {
                    bucket.push(doll);
                }

                board[j][column] = 0;
                break;
            }
        }

    }


    return answer * 2;
}

File: wooyoo/Programmers/2019_카카오_개발자_겨울_인턴십/PR_크레인_인형뽑기_게임.js
function take_doll(board, x, depth) {
  let doll_type = 0;
  for (let i = 0; i < depth; i++) {
    if (board[i][x] != 0) {
      doll_type = board[i][x];
      board[i][x] = 0;
      return doll_type;
    }
  }
  return doll_type;
}

function solution(board, moves) {
  var answer = 0;
  let bucket = [];

  let new_doll;
  for (let i of moves) {
    //   console.log(board);
    new_doll = take_doll(board, i - 1, board.length);
    if (new_doll !== 0) {
      if (bucket[0] === new_doll) {
        bucket.shift();
        answer++;
      } else {
        bucket.unshift(new_doll);
      }
    }
    //   console.log(bucket);
  }
  return answer * 2;
}
